src/python/pants/backend/experimental/python/lockfile_metadata.py

```diff
@@ -17,51 +18,97 @@
 @dataclass
 class LockfileMetadata:
     invalidation_digest: str | None
+    valid_interpreter_constraints: InterpreterConstraints | None
```

Comment:
"valid" is throwing me off because it's making me wonder what "invalid" would be. Maybe something `interpreter_constraints`? Given that it's an attribute of `LockfileMetadata`, I think it's clear enough it's the "ICs used when generating this lockfile."

If that's not clear enough, could say something more verbose like `interpreter_constraints_during_generation`?

Reply:
The aim here is to describe the interpreter constraints _for which this lockfile is valid_, so it's valid in the same way as we have an (in)validation digest

Reply:
Hm although note that the ICs were possibly only valid at the time of generating the lockfile, and might not be valid anymore if the user has changed their ICs. 

That's I think what most throws me off: including `valid_` is claiming more than what this is: simply the ICs that were used when generating the lockfile. At one time, they were valid, and they still might be. But that isn't determined until comparing against the context's ICs.

I recommend simply `LockfileMetadata.interpreter_constraints`, representing the ICs used when that lockfile was generated.

Reply:
I don't think `interpreter_constraints` explains the point of having them in the file. The interpreter constraints in the file are always going to be the ones under which the lockfile is valid -- if the user changes their ICs, then the _user's_ ICs are invalid until a lockfile is created, not the other way around.

Reply:
> if the user changes their ICs, then the user's ICs are invalid until a lockfile is created, not the other way around.

From a modeling perspective, I think it's the exact way around. If you updated your ICs, you did so for a reason. And as a consequence, the lockfile is now stale and needs to be regenerated. The issue isn't your new ICs, it's that the lockfile is now stale.

We don't want to imply with 'valid_interpreter_constraints' that you must constrain yourself to the ICs in your lockfiles, that only those are valid. If you want to change the ICs, go for it, only regenerate the lockfile.

Reply:
OK, I'm going with `valid_for_interpreter_constraints` and `requirements_invalidation_digest`

Reply:
I like `valid_for_interpreter_constraints`! Thanks, Chris.